File: go/crud/event_metadata.go
package crud

import (
	"github.com/fraym/freym-api/go/proto/crud/deliverypb"
)

type EventMetadata struct {
	CorrelationId string
	CausationId   string
	UserId        string
	Target        deliverypb.DeploymentTarget
}

func (m *EventMetadata) getProtobufEventMetadata() *deliverypb.EventMetadata {
	if m == nil {
		return nil
	}

	return deliverypb.EventMetadata_builder{
		CorrelationId: m.CorrelationId,
		CausationId:   m.CausationId,
		Target:        m.Target,
		UserId:        m.UserId,
	}.Build()
}

File: js/crud/client/eventMetadata.ts
import {
    EventMetadata as CrudEventMetadata,
    DeploymentTarget,
} from "@fraym/proto/dist/index.freym.crud.delivery";

export interface EventMetadata {
    correlationId: string;
    causationId: string;
}

export const fillMetadataWithDefaults = (
    metadata: Partial<EventMetadata> | null,
    target: DeploymentTarget
): CrudEventMetadata => {
    if (!metadata) {
        return {
            causationId: "",
            correlationId: "",
            target,
        };
    }

    return {
        causationId: metadata.causationId || "",
        correlationId: metadata.correlationId || "",
        target,
    };
};
